tonic/src/transport/service/timeout.rs

```diff
@@ -0,0 +1,321 @@
+                     Using server timeout of: {:?}",
+                    header, server, shorter_duration,
+                );
+                Some(shorter_duration)
+            }
+        };
+
+        let inner_future = self.inner.call(req);
+        future::TimeoutFuture::new(inner_future, timeout_duration)
+    }
+}
+
+/// Utility methods for parsing the gRPC headers
+mod headers {
+    use http::{HeaderMap, HeaderValue};
+    use std::time::Duration;
+
+    const GRPC_TIMEOUT_HEADER: &str = "grpc-timeout";
+
+    const SECONDS_IN_HOUR: u64 = 60 * 60;
+    const SECONDS_IN_MINUTE: u64 = 60;
+
+    /// Tries to parse the `grpc-timeout` header if it is present. If we fail to parse, returns
+    /// the value we attempted to parse.
+    ///
+    /// Follows the [gRPC over HTTP2 spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md).
+    pub(crate) fn try_parse_grpc_timeout(
+        headers: &HeaderMap<HeaderValue>,
+    ) -> Result<Option<Duration>, &HeaderValue> {
+        match headers.get(GRPC_TIMEOUT_HEADER) {
+            Some(val) => {
+                let str_val = val.to_str().map_err(|_| val)?;
+                let (timeout_value, timeout_unit) = try_split_last(str_val).map_err(|_| val)?;
+
+                // gRPC spec specifies `TimeoutValue` will be at most 8 digits
+                // Caping this at 8 digits also prevents integer overflow from ever occurring
+                if timeout_value.len() > 8 {
+                    return Err(val);
+                }
+
+                let timeout_value: u64 = timeout_value.parse().map_err(|_| val)?;
+
+                let duration = match timeout_unit {
+                    // Hours
+                    "H" => Duration::from_secs(timeout_value * SECONDS_IN_HOUR),
+                    // Minutes
+                    "M" => Duration::from_secs(timeout_value * SECONDS_IN_MINUTE),
+                    // Seconds
+                    "S" => Duration::from_secs(timeout_value),
+                    // Milliseconds
+                    "m" => Duration::from_millis(timeout_value),
+                    // Microseconds
+                    "u" => Duration::from_micros(timeout_value),
+                    // Nanoseconds
+                    "n" => Duration::from_nanos(timeout_value),
+                    _ => return Err(val),
+                };
+
+                Ok(Some(duration))
+            }
+            None => Ok(None),
+        }
+    }
+
+    /// Tries to split the last character of the string, from the rest of the string,
+    /// returning (rest_of_string, last_char), if successful.
+    ///
+    /// `str.split_at(...)` panics if we're not on a UTF-8 code point boundary. This
+    /// should never happen in practice because the `grpc-timeout` header should be only
+    /// ASCII characters.
+    fn try_split_last(val: &str) -> Result<(&str, &str), &str> {
+        std::panic::catch_unwind(|| val.split_at(val.len() - 1)).map_err(|_| val)
```

Comment:
I realized, after writing a unit test, that `HeaderValue` already prevents non-ASCII characters 🎉 so this extra `catch_unwind(|| ...)` probably isn't necessary, but will leave it in for now

Reply:
I guess `str.split_at(...)` should be fine since `HeaderValue::to_str` checks for ASCII as you mention.